Attribute plan modification

kmoe opened the pull request:
Closes #34 

Design doc: #64 

Needs #107

This PR implements `schema.Attribute.PlanModifiers` and the `RequiresReplace` and `RequiresReplaceIf` helpers, as described in the design doc.

Some more design points that arose during implementation:
1. When multiple `AttributePlanModifiers` are specified, should the `req.AttributePlan` value supplied to each be the `resp.AttributePlan` value from the previous modifier? This seems the intuitive behaviour (see `chained_modifiers` test for example).
2. If an error diagnostic is returned at some point during attribute plan modification, should `RequiresReplace` and the planned state still be populated in the server response? This PR assumes so, and that it is Core's responsibility to decide what to do in this case. This also makes it easier to write deterministic tests, as there is otherwise a race condition between early return on certain errors, and RequiresReplace being populated.

paddycarver commented:
> When multiple AttributePlanModifiers are specified, should the req.AttributePlan value supplied to each be the resp.AttributePlan value from the previous modifier? This seems the intuitive behaviour (see chained_modifiers test for example).

This matches my expectations. I'm not sure how we'd let them all operate on the request; what if two or more modified the same field in different ways? Which result would we use?

> If an error diagnostic is returned at some point during attribute plan modification, should RequiresReplace and the planned state still be populated in the server response? This PR assumes so, and that it is Core's responsibility to decide what to do in this case. This also makes it easier to write deterministic tests, as there is otherwise a race condition between early return on certain errors, and RequiresReplace being populated.

This also matches my expectations, and my mental model of how plans works tells me it doesn't matter. Unlike applies, plans shouldn't get persisted in the face of an error, I believe, so if an error diagnostic is being returned, I think the plan is just going to get discarded, anyways. So this seems like reasonable behavior that won't have any external negative impacts, as far as I know.

paddycarver commented on tfsdk/schema.go:
```diff
@@ -208,3 +210,189 @@ func (s Schema) validate(ctx context.Context, req ValidateSchemaRequest, resp *V
+				}
+
+				for idx := range l.Elems {
+					for nestedName, nestedAttr := range a.Attributes.GetAttributes() {
```
I think we want this to be recursive; right now, it looks like you only have one level of nesting supported. So a nested attribute inside a nested attribute inside a schema wouldn't have its plan modifiers run, as best I can tell.

If we made this method into a function that accepted a `map[string]Attribute` and an attribute path, we could just recursively call it for nested attributes:

```go
func modifyAttributesPlans(ctx context.Context, attrs map[string]Attribute, path *tftypes.AttributePath, req ModifySchemaPlanRequest, resp *ModifySchemaPlanResponse) {
```

Then this section could be something like

```go
for idx := range l.Elems {
  modifyAttributesPlans(ctx, a.Attributes.GetAttributes(), path.WithElementKeyInt(idx), req, resp)
  if diagnostics.DiagsHasErrors(resp.Diagnostics) {
    return
  }
}
```

Or at least that's what my vague imagining is suggesting it could look like :D

paddycarver replied:
This could also (maybe?) be resolved by doing the recursion on the `Attribute.modifyPlan` method, above.

paddycarver commented on tfsdk/serve_provider_test.go:
```diff
@@ -431,6 +431,7 @@ func (t *testServeProvider) GetResources(_ context.Context) (map[string]Resource
 	return map[string]ResourceType{
 		"test_one":               testServeResourceTypeOne{},
 		"test_two":               testServeResourceTypeTwo{},
+		"test_three":             testServeResourceTypeThree{},
```
Is it worth giving this a name like the validators below, to make it clear it's testing plan modifiers? 🤔

kmoe commented:
> what if two or more modified the same field in different ways? Which result would we use?

`Attribute.PlanModifiers` is a slice of `AttributePlanModifiers`, executed in order. Each one can modify only the planned value for that attribute. Given this I think it's clear that if several plan modifiers on one attribute modify its value, the last modified value should be used.

Some complexity comes with nested attributes - currently it's possible to define `PlanModifiers` at several levels of the nested attribute hierarchy, e.g.:

```go
Schema{
	Attributes: map[string]Attribute{
		"disks": {
			Optional: true,
			Computed: true,
			PlanModifiers: ..., // modifies "disks"
			Attributes: ListNestedAttributes(map[string]Attribute{
				"name": {
					Required: true,
					Type:     types.StringType,
					PlanModifiers: ..., // modifies "disks[0].name" etc
				},
				"size_gb": {
					Required: true,
					Type:     types.NumberType,
				},
				"boot": {
					Required: true,
					Type:     types.BoolType,
				},
			}, ListNestedAttributesOptions{}),
		},
	},
}
```

Currently, the top-level PlanModifiers are run first, followed by any modifiers on nested attributes.

This at least needs to be noted in the GoDoc and a test case added, so I'll do that.

If we like we could also prohibit PlanModifiers on Attributes without a Type, but I don't see a need to do this as long as the order is unambiguous and clearly documented.

paddycarver commented on tfsdk/attribute.go:
```diff
@@ -369,3 +376,42 @@ func (a Attribute) validate(ctx context.Context, req ValidateAttributeRequest, r
+	for _, planModifier := range a.PlanModifiers {
+		planModifier.Modify(ctx, modifyReq, resp)
+		modifyReq.AttributePlan = resp.AttributePlan
+	}
```
Is it worth making this recursive, like `Attribute.validate` is, so it goes through all the nested attributes of this attribute and calls their `modifyPlan` method, too?

paddycarver commented on tfsdk/serve_test.go:
```diff
@@ -1874,14 +1899,318 @@ func TestServerPlanResourceChange(t *testing.T) {
 		name, tc := name, tc
 
 		t.Run(name, func(t *testing.T) {
-			t.Parallel()
```
Hmm, why did this need to stop running in parallel? 🤔 Would love a comment here explaining why it can't run in parallel. That makes me mildly worried about race conditions.

kmoe replied:
Thanks, I don't think this removal is needed any more since I got to the bottom of the race conditions. Will restore it.

paddycarver commented on tfsdk/serve_test.go:
```diff
@@ -1940,7 +2269,9 @@ func TestServerPlanResourceChange(t *testing.T) {
 				return
 			}
-			if diff := cmp.Diff(got.RequiresReplace, tc.expectedRequiresReplace, cmpopts.EquateEmpty()); diff != "" {
+			if diff := cmp.Diff(got.RequiresReplace, tc.expectedRequiresReplace, cmpopts.EquateEmpty(), cmpopts.SortSlices(func(x, y *tftypes.AttributePath) bool {
```
Hmm, rather than new modifiers on cmp.Diff, does implementing an `Equal` method allow us to make this work without the cmpopts?

kmoe replied:
Yes, that would work. We could do one of the following:

1. Define a `type  RequiresReplacePaths []*tftypes.AttributePath` in the framework with an `Equal` method that ignores ordering
2. Do something similar but in plugin-go/tftypes itself.

However, I think it might be better just to have the framework server sort them so the response is deterministic, rather than ignoring the change of order during testing. Added 583f552d0530a5c7d4ddff4b551aa3da32eb8cd3 and a8991d73b60056734455d5afaded3611c1204fa0 for this purpose.

paddycarver commented on tfsdk/serve_test.go:
```diff
@@ -3088,8 +3419,6 @@ func TestServerApplyResourceChange(t *testing.T) {
 			}
 			if tc.config.Type() != nil {
 				if diff := cmp.Diff(s.applyResourceChangeConfigValue, tc.config); diff != "" {
-					t.Errorf("Unexpected diff in config (+wanted, -got): %s", diff)
```
This slightly concerns me; are we just not using this `if` check anymore?

kmoe replied:
Wow, I have no idea where this completely unrelated change came from. Thanks for spotting it - reverted

paddycarver reviewed (changes requested):
A few comments, mostly about some testing stuff and about recursion.

paddycarver commented:
Also, some merge conflicts to resolve, I believe.

paddycarver commented:
> `Attribute.PlanModifiers` is a slice of `AttributePlanModifiers`, executed in order. Each one can modify only the planned value for that attribute. Given this I think it's clear that if several plan modifiers on one attribute modify its value, the last modified value should be used.

Yeah, sorry, I think we're getting at the same thing here. I was trying to say I don't know how it would work if we _didn't_ take that approach.

> Some complexity comes with nested attributes - currently it's possible to define `PlanModifiers` at several levels of the nested attribute hierarchy, e.g.:
> 
> [snip]
> 
> Currently, the top-level PlanModifiers are run first, followed by any modifiers on nested attributes.
> 
> This at least needs to be noted in the GoDoc and a test case added, so I'll do that.
> 
> If we like we could also prohibit PlanModifiers on Attributes without a Type, but I don't see a need to do this as long as the order is unambiguous and clearly documented.

Yeah, I'm fine with just "top-level goes first, nested ones follow", that seems like a really reasonable approach and I think the complexity is worth the extra power.

appilon commented:
Would this feature allow a provider to specify defaults for optional fields?

kmoe commented on tfsdk/attribute.go:
```diff
@@ -73,6 +73,13 @@ type Attribute struct {
+	// PlanModifiers defines a sequence of modifiers for this attribute at
+	// plan time.
+	// Please note that plan modification only applies to resources, not
+	// data sources. Setting PlanModifiers on a data source attribute will
+	// have no effect.
```
Opened https://github.com/hashicorp/terraform-plugin-framework/issues/113 to track `InternalValidate` options. I think we might be able to get away without it but feel free to add anything else that might be appropriate to that issue.

kmoe commented:
@appilon Yes, thanks for raising this. Adding a test case to show this working as expected.

A helper func like this makes sense I think:
```go
Schema{
	Attributes: map[string]Attribute{
		"size": {
			Optional: true,
			PlanModifiers: []AttributePlanModifier{AttributeDefault(val)},
		},
	},
}
```
where we have either `func AttributeDefault(v interface{}) AttributePlanModifier` or `func AttributeDefault(v attr.Value) AttributePlanModifier`. Similar helper makes sense for a `DefaultFunc` equivalent.

Will follow up with a further PR for any such helpers. Also considering putting `RequiresReplace` and any other helpers in their own package so they are namespaced.

kmoe commented:
Implementation of full recursion of nested attributes works, but the tests are currently running into https://github.com/hashicorp/terraform-plugin-framework/issues/112 as `tftypes.WalkAttributePath` (called in `Plan.terraformValueAtPath()`) cannot handle a nil `tftypes.Object`. I suspect this is a plugin-go fix and will write it up there.

tckb commented:
+1 waiting for this PR merged - I have the same issue as mentioned in #112

kmoe commented:
In fact I think this doesn't require a plugin-go change (see comment on #112).

Changes since review:
 - fixed recursion, so attribute plan modifiers run for arbitrarily nested attributes
 - added sorting and deduplication to `resp.RequiresReplace` slice so it is entirely deterministic
 - added test case for setting default attribute values
 - added early return after first plan modifier that errors, instead of continuing with possibly invalid plan values

Ready for review.

paddycarver commented on tfsdk/attribute_plan_modification.go:
```diff
@@ -0,0 +1,204 @@
+// RequiresReplace returns an AttributePlanModifier specifying the attribute as
+// requiring replacement. This behaviour is identical to the ForceNew behaviour
+// in terraform-plugin-sdk.
+func RequiresReplace() AttributePlanModifier {
```
nitpick: if we're going to have a constructor, I feel like the underlying type could be unexported to limit our compatibility surface. If we're going to export the underlying type, I don't know that we need a constructor. 🤔 

(This should not block merge, we can iterate later.)

paddycarver commented on tfsdk/plan.go:
```diff
@@ -167,6 +167,7 @@ func (p *Plan) SetAttribute(ctx context.Context, path *tftypes.AttributePath, va
 func (p Plan) terraformValueAtPath(path *tftypes.AttributePath) (tftypes.Value, error) {
 	rawValue, remaining, err := tftypes.WalkAttributePath(p.Raw, path)
 	if err != nil {
+
```
nit: extraneous whitespace.

(Should not block merge)

paddycarver reviewed (approved):
Awesome work! This looks great to me.

bflad commented on tfsdk/schema.go:
```diff
@@ -208,3 +210,120 @@ func (s Schema) validate(ctx context.Context, req ValidateSchemaRequest, resp *V
+			resp.RequiresReplace = append(resp.RequiresReplace, attrPath)
+		}
+
+		setAttrDiags := resp.Plan.SetAttribute(ctx, attrPath, nestedAttrResp.AttributePlan)
+		resp.Diagnostics = append(resp.Diagnostics, setAttrDiags...)
+		if diagnostics.DiagsHasErrors(setAttrDiags) {
+			return
+		}
+		resp.Diagnostics = nestedAttrResp.Diagnostics
```
Is this last line expected? I _think_ its just extraneous since they should point to the same address, but it could be  confusing as it looks like it might be removing `SetAttribute` diagnostics.
